common/flatpak-run-wayland.c

```diff
@@ -0,0 +1,218 @@
+#include "flatpak-run-wayland-private.h"
+
+#ifdef ENABLE_WAYLAND_SECURITY_CONTEXT
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <wayland-client.h>
+#include "security-context-v1-protocol.h"
+#endif
+
+#include "flatpak-utils-private.h"
+
+#ifdef ENABLE_WAYLAND_SECURITY_CONTEXT
+
+static void registry_handle_global (void *data, struct wl_registry *registry,
+                                    uint32_t name, const char *interface,
+                                    uint32_t version)
+{
+  struct wp_security_context_manager_v1 **out = data;
+
+  if (strcmp (interface, wp_security_context_manager_v1_interface.name) == 0) {
+    *out = wl_registry_bind (registry, name,
+                             &wp_security_context_manager_v1_interface, 1);
+  }
+}
+
+static void registry_handle_global_remove (void *data,
+                                           struct wl_registry *registry,
+                                           uint32_t name)
+{
+  /* no-op */
+}
+
+static const struct wl_registry_listener registry_listener = {
+  .global = registry_handle_global,
+  .global_remove = registry_handle_global_remove,
+};
+
+static char *
+create_wl_socket (char *template)
+{
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  g_autofree char *proxy_socket_dir = g_build_filename (user_runtime_dir, ".flatpak/wl", NULL);
+  g_autofree char *proxy_socket = g_build_filename (proxy_socket_dir, template, NULL);
+  int fd;
+
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, proxy_socket_dir, 0755, NULL, NULL))
+    return NULL;
+
+  fd = g_mkstemp (proxy_socket);
+  if (fd == -1)
+    return NULL;
+
+  close (fd);
+
+  return g_steal_pointer (&proxy_socket);
+}
+
+static gboolean
+flatpak_run_add_wayland_security_context_args (FlatpakBwrap *bwrap,
+                                               const char   *app_id,
+                                               const char   *instance_id,
+                                               gboolean     *available_out)
+{
+  gboolean res = FALSE;
+  struct wl_display *display;
+  struct wl_registry *registry;
+  struct wp_security_context_manager_v1 *security_context_manager = NULL;
+  struct wp_security_context_v1 *security_context;
+  struct sockaddr_un sockaddr = {0};
+  g_autofree char *socket_path = NULL;
+  int listen_fd = -1, sync_fd, ret;
+
+  *available_out = TRUE;
+
+  display = wl_display_connect (NULL);
+  if (!display)
+    return FALSE;
+
+  registry = wl_display_get_registry (display);
+  wl_registry_add_listener (registry, &registry_listener,
+                            &security_context_manager);
+  ret = wl_display_roundtrip (display);
+  wl_registry_destroy (registry);
+  if (ret < 0)
+    goto out;
+
+  if (!security_context_manager) {
+    *available_out = FALSE;
+    goto out;
+  }
+
+  socket_path = create_wl_socket ("wayland-XXXXXX");
+  if (!socket_path)
+    goto out;
+
+  unlink (socket_path);
+
+  listen_fd = socket (AF_UNIX, SOCK_STREAM, 0);
```

Comment:
So, we're leaking listen_fd in the success case. This is generally not too bad, but if we want to close it, where can we do so safely? I guess after `wp_security_context_manager_v1_create_listener` and `wl_display_roundtrip` have been issued the fd is on the other end and will stay open.

Reply:
we're not. never mind.

Reply:
This has been discussed in https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/232#note_1991868